volatility3/framework/symbols/linux/utilities/module_extract.py

```diff
@@ -0,0 +1,891 @@
+                context, vmlinux_name, module.sect_attrs.grp
+            )
+
+        if num_sections > 1024 or num_sections == 0:
+            vollog.debug(
+                f"Invalid number of sections ({num_sections}) for module at offset {module.vol.offset:#x}"
+            )
+            return None
+
+        vmlinux = context.modules[vmlinux_name]
+
+        # This is declared as a zero sized array, so we create ourselves
+        attribute_type = module.sect_attrs.attrs.vol.subtype
+
+        sect_array = vmlinux.object(
+            object_type="array",
+            subtype=attribute_type,
+            offset=module.sect_attrs.attrs.vol.offset,
+            count=num_sections,
+            absolute=True,
+        )
+
+        sections: Dict[int, str] = {}
+
+        # for each section, gather its name and address
+        for index, section in enumerate(sect_array):
+            name = section.get_name()
+
+            sections[section.address] = name
+
+        return sections
+
+    @classmethod
+    def _parse_sections(
+        cls,
+        context: interfaces.context.ContextInterface,
+        vmlinux_name: str,
+        module: extensions.module,
+    ) -> Optional[Tuple[List, int, int]]:
+        """
+        This function first parses the sections as maintained by the kernel
+        It then orders the sections by load address, and then gathers the data of each section
+        We also track the file_offset to correctly have alignment in the output file
+
+        .symtab requires special handling as its so broken in memory as described in `_fix_sym_table`
+        The data of .strtab is read directly off the module structure and not its section
+        as the section from the original module has no meaning after loading as the kernel does not reference it.
+        """
+        original_sections = cls._enumerate_original_sections(
+            context, vmlinux_name, module
+        )
+        if original_sections is None:
+            return None
+
+        kernel = context.modules[vmlinux_name]
+        kernel_layer = context.layers[kernel.layer_name]
+
+        if symbols.symbol_table_is_64bit(context, kernel.symbol_table_name):
+            sym_type = "Elf64_Sym"
+            elf_hdr_type = "Elf64_Ehdr"
+            st_fmt = "<Q"
+        else:
+            sym_type = "Elf32_Sym"
+            elf_hdr_type = "Elf32_Ehdr"
+            st_fmt = "<I"
+
+        # At this point, we have the sections starting addresses and names,
+        # but the kernel does not track the size
+        # To recover the size, we sort by address and then use the next section as the boundary to calculate size
+        # .symtab (the symbol table) and .strtab (the strings table) require special handling.
+        # All others can be read with padding
+
+        # get the addresses in sorted order, can index into `original_sections` for names
+        sorted_addresses = sorted(original_sections)
```

Comment:
This works, but it's a little confusing that it converts it to a list and then sorts the list.  Could we add `.keys()` to make it more explicit what's going on please?

Reply:
good catch. I used in keys() for that reason in other plugins.